src/forder.c

```diff
@@ -1420,3 +1465,249 @@ SEXP binary(SEXP x)
   return ans;
 }
 
+// all(x==1L)
+static bool all1(SEXP x) {
+  if (!isInteger(x))
+    error("internal error: all1 got non-integer"); // # nocov
+  int *xp = INTEGER(x);
+  for (int i=0; i<LENGTH(x); ++i) if (xp[i] != 1) return false;
+  return true;
+}
+
+// identical(cols, head(chmatch(key(x), names(x)), length(cols)))
+bool colsKeyHead(SEXP x, SEXP cols) {
+  if (!isNewList(x))
+    error("internal error: 'x' must be a list"); // # nocov
+  if (!isInteger(cols))
+    error("internal error: 'cols' must be an integer"); // # nocov
+  SEXP key = getAttrib(x, sym_sorted);
+  if (isNull(key) || (length(key) < length(cols)))
+    return false;
+  SEXP names =  getAttrib(x, R_NamesSymbol);
```

Comment:
do we need to `PROTECT()` here?

Reply:
I doubt that names would be gc'ed, `x` would have to be gc'ed as well and `x` is protected. I think if `rchk` is fine then we can ignore.